In [1]:
import numpy as np
import pandas as pd

# Exploration

## Read JSON of pipeline data

In [55]:
df = pd.read_json('./info from WorkStep/candidate_funnel_2022_08.json')

In [56]:
df.shape

(40935, 15)

In [57]:
df['position_id'].nunique()

2670

In [58]:
df = df.fillna(0)

In [59]:
needs_ints = ['count_started', 'count_submitted', 'count_approved', 'count_reviewed',
       'count_expired', 'count_withdrawn', 'count_deactivated',
       'count_position_closed', 'count_invited_to_interview', 'count_hired']
df[needs_ints] = df[needs_ints].astype(int)

In [52]:
for col in needs_ints:
    df[col].apply(lambda x: int(x))

There are 40k+ entries in the dataset, but only 2670 positions.

Let's look at a sample position which was hired for, to see what the pipeline looks like

In [60]:
df8049247818 = df[(df['company_id'] == 69615) & (df['position_id'] == 8049247818)].head(50)

In [61]:
df8049247818

company_id position_role_type  position_id                      date  \
36       69615          WAREHOUSE   8049247818 2022-08-01 00:00:00+00:00   
37       69615          WAREHOUSE   8049247818 2022-08-02 00:00:00+00:00   
38       69615          WAREHOUSE   8049247818 2022-08-03 00:00:00+00:00   
39       69615          WAREHOUSE   8049247818 2022-08-08 00:00:00+00:00   
40       69615          WAREHOUSE   8049247818 2022-08-11 00:00:00+00:00   
41       69615          WAREHOUSE   8049247818 2022-08-12 00:00:00+00:00   
42       69615          WAREHOUSE   8049247818 2022-08-14 00:00:00+00:00   
43       69615          WAREHOUSE   8049247818 2022-08-15 00:00:00+00:00   
44       69615          WAREHOUSE   8049247818 2022-08-16 00:00:00+00:00   
45       69615          WAREHOUSE   8049247818 2022-08-18 00:00:00+00:00   
46       69615          WAREHOUSE   8049247818 2022-08-21 00:00:00+00:00   
47       69615          WAREHOUSE   8049247818 2022-08-22 00:00:00+00:00   
48       69615          WAREHOUSE   8049247818 2022-08-29 00:00:00+00:00   

       spend  count_started  count_submitted  count_approved  count_reviewed  \
36  0.005527              2                1               1               0   
37  0.005857              0                0               0               1   
38  0.046157              1                1               1               0   
39  0.011225              2                1               1               0   
40  0.000549              1                1               1               3   
41  0.008016              1                0               0               0   
42  0.011640              1                2               2               0   
43  0.023024              3                0               0               0   
44  0.023866              1                1               1               2   
45  0.004319              1                1               1               0   
46  0.004905              1                0               0               0   
47  0.016032              1                1               0               2   
48  0.000000              0                0               0               0   

    count_expired  count_withdrawn  count_deactivated  count_position_closed  \
36              0                0                  0                      0   
37              0                0                  0                      0   
38              0                0                  0                      0   
39              0                0                  0                      0   
40              0                0                  0                      0   
41              0                0                  0                      0   
42              0                0                  0                      0   
43              0                0                  0                      0   
44              0                0                  0                      0   
45              0                0                  0                      0   
46              0                0                  0                      0   
47              0                0                  0                      0   
48              0                0                  0                      0   

    count_invited_to_interview  count_hired  
36                           0            0  
37                           1            0  
38                           0            0  
39                           0            0  
40                           3            0  
41                           0            0  
42                           0            0  
43                           0            0  
44                           1            0  
45                           0            0  
46                           0            0  
47                           1            0  
48                           0            1

In [62]:
summable_cols = ['spend', 'count_started', 'count_submitted', 'count_approved', 'count_reviewed',
       'count_expired', 'count_withdrawn', 'count_deactivated',
       'count_position_closed', 'count_invited_to_interview', 'count_hired']

In [63]:
df8049247818[summable_cols].sum()

spend                          0.161117
count_started                 15.000000
count_submitted                9.000000
count_approved                 8.000000
count_reviewed                 8.000000
count_expired                  0.000000
count_withdrawn                0.000000
count_deactivated              0.000000
count_position_closed          0.000000
count_invited_to_interview     6.000000
count_hired                    1.000000
dtype: float64

According to this data:

* 15 people began applications
* 9 completed them
* 8 passed initial screening
* 8 were reviewed by the HM
* 6 invited to interview
* 1 hired

## Group by position ID

In [65]:
df_positions = df.groupby('position_id', as_index=False)[summable_cols].sum()
df_positions[df_positions['count_hired'] >= 1]

position_id     spend  count_started  count_submitted  count_approved  \
2       101331110  5.786332            797              496             381   
3       101926072  0.041716             11                8               7   
8       103393641  1.173819            226              169             126   
10      103630930  0.139777            100               81              79   
11      104081540  1.548433            294              181             148   
...           ...       ...            ...              ...             ...   
2651   9178975289  0.846082            176              131              99   
2652   9194306483  0.872192            133               92              86   
2654   9196561636  0.769348             57               33              26   
2667   9905840884  0.788126            159              100              95   
2669   9997425444  0.137776             19               14               8   

      count_reviewed  count_expired  count_withdrawn  count_deactivated  \
2                381              0                0                  0   
3                  7              0                0                  0   
8                126              0                0                  0   
10                79              0                0                  0   
11               111             36                0                  0   
...              ...            ...              ...                ...   
2651              99              0                0                  0   
2652              82              4                0                  0   
2654              21              5                0                  0   
2667              88              7                0                  0   
2669               8              0                0                  0   

      count_position_closed  count_invited_to_interview  count_hired  
2                         0                         185            1  
3                         0                           7            1  
8                         0                         126            1  
10                        0                          45            1  
11                        1                          64            2  
...                     ...                         ...          ...  
2651                      0                          99            1  
2652                      0                          23            2  
2654                      0                          18            1  
2667                      0                           3            1  
2669                      0                           8            1  

[515 rows x 12 columns]

* Of the 2670 positions in the dataset, 2155 ended in no hire and only 515 ended with a hire!

In [71]:
df_positions['percent_submitted'] = round(df_positions['count_submitted'] / df_positions['count_started'] * 100.0, 2)
df_positions['percent_approved'] = round(df_positions['count_approved'] / df_positions['count_submitted'] * 100.0, 2)
df_positions['percent_reviewed'] = round(df_positions['count_reviewed'] / df_positions['count_approved'] * 100.0, 2)
df_positions['percent_interviewed'] = round(df_positions['count_invited_to_interview'] / df_positions['count_reviewed'] * 100.0, 2)
df_positions['percent_hired'] = round(df_positions['count_hired'] / df_positions['count_invited_to_interview'] * 100.0, 2)

In [72]:
df_positions

position_id     spend  count_started  count_submitted  count_approved  \
0       100286000  0.595761             64               40              36   
1       100778742  0.396039             41               27              15   
2       101331110  5.786332            797              496             381   
3       101926072  0.041716             11                8               7   
4       101998387  0.690089             42               21              12   
...           ...       ...            ...              ...             ...   
2665   9696000534  0.074183             11                6               3   
2666   9868359373  0.236643             27               21              18   
2667   9905840884  0.788126            159              100              95   
2668   9967558006  0.005686              3                3               3   
2669   9997425444  0.137776             19               14               8   

      count_reviewed  count_expired  count_withdrawn  count_deactivated  \
0                 36              0                0                  0   
1                  5             10                0                  0   
2                381              0                0                  0   
3                  7              0                0                  0   
4                 12              0                0                  0   
...              ...            ...              ...                ...   
2665               3              0                0                  0   
2666              18              0                0                  0   
2667              88              7                0                  0   
2668               3              0                0                  0   
2669               8              0                0                  0   

      count_position_closed  count_invited_to_interview  count_hired  \
0                         0                           1            0   
1                         0                           5            0   
2                         0                         185            1   
3                         0                           7            1   
4                         0                           7            0   
...                     ...                         ...          ...   
2665                      0                           2            0   
2666                      0                           4            0   
2667                      0                           3            1   
2668                      0                           3            0   
2669                      0                           8            1   

      percent_submitted  percent_approved  percent_reviewed  \
0                 62.50             90.00            100.00   
1                 65.85             55.56             33.33   
2                 62.23             76.81            100.00   
3                 72.73             87.50            100.00   
4                 50.00             57.14            100.00   
...                 ...               ...               ...   
2665              54.55             50.00            100.00   
2666              77.78             85.71            100.00   
2667              62.89             95.00             92.63   
2668             100.00            100.00            100.00   
2669              73.68             57.14            100.00   

      percent_interviewed  percent_hired  
0                    2.78           0.00  
1                  100.00           0.00  
2                   48.56           0.54  
3                  100.00          14.29  
4                   58.33           0.00  
...                   ...            ...  
2665                66.67           0.00  
2666                22.22           0.00  
2667                 3.41          33.33  
2668               100.00           0.00  
2669               100.00          12.50  

[2670 rows x 17 c

# 1. Performance

## Group by company ID

In [75]:
df_gb_company = df.groupby('company_id', as_index=False)[summable_cols].sum()
df_gb_company['spend_per_hire_USD'] = df_gb_company['spend'] / df_gb_company['count_hired']
df_gb_company.sort_values(by='spend_per_hire_USD', inplace=True, ascending=True)

In [76]:
df_gb_company.head(10)

company_id      spend  count_started  count_submitted  count_approved  \
80       217743   0.447858             78               57              42   
118      629687  14.589868           1796             1339            1124   
5         14221   0.360912             92               71              62   
2         12155   0.995217             89               57              46   
12        20805   1.003038            218              170             145   
20        27712   4.752413            742              502             349   
102      431776   0.208275             17                9               8   
25        34548   4.421479            660              438             346   
91       329237   1.483486            128               66              54   
127      731460   0.254286             46               34              24   

     count_reviewed  count_expired  count_withdrawn  count_deactivated  \
80               42              0                0                  0   
118            1084             38                1                  0   
5                57              1                0                  0   
2                46              0                0                  0   
12              136              2                0                  0   
20              323             24                0                  0   
102               8              0                0                  0   
25              344              0                1                  0   
91               54              0                0                  0   
127              24              0                0                  0   

     count_position_closed  count_invited_to_interview  count_hired  \
80                       0                           8            3   
118                      1                         945           85   
5                        4                          52            2   
2                        0                          46            5   
12                       7                          51            5   
20                       0                         262           23   
102                      0                           7            1   
25                       1                         326           18   
91                       0                          45            6   
127                      0                           3            1   

     spend_per_hire_USD  
80             0.149286  
118            0.171646  
5              0.180456  
2              0.199043  
12             0.200608  
20             0.206627  
102            0.208275  
25             0.245638  
91             0.247248  
127            0.254286

In [81]:
opp_companies = df_gb_company[df_gb_company['count_hired'] > 0].sort_values(by='spend_per_hire_USD', ascending=False)
opp_companies.head(10)

company_id      spend  count_started  count_submitted  count_approved  \
39        54042  12.626533           1769             1034             621   
97       403553  27.676070           5543             3940            3189   
92       346676   3.980185            352              210             134   
54        73394   3.732378            932              694             561   
140      868835   3.561146            571              324             198   
98       407848   3.268647            459              278             139   
45        58946   6.225625            642              415             329   
106      468091   6.089142            688              446             277   
29        41679  64.157676           7809             5186            3962   
108      498096   7.632267           1472             1016             630   

     count_reviewed  count_expired  count_withdrawn  count_deactivated  \
39              607             14                0                  0   
97             2762             74                1                  0   
92               96             37                1                  0   
54              345            216                0                  0   
140              51            115                0                  0   
98              123             14                0                  0   
45              213            115                0                  0   
106             267              8                1                  0   
29             3961              0                0                  0   
108             613             17                0                  0   

     count_position_closed  count_invited_to_interview  count_hired  \
39                       0                         124            1   
97                     352                         599            6   
92                       0                          21            1   
54                       0                          33            1   
140                     32                          32            1   
98                       1                          20            1   
45                       0                         175            2   
106                      0                         124            2   
29                       1                        3922           24   
108                      0                          39            3   

     spend_per_hire_USD  
39            12.626533  
97             4.612678  
92             3.980185  
54             3.732378  
140            3.561146  
98             3.268647  
45             3.112813  
106            3.044571  
29             2.673237  
108            2.544089

## Group by role type

In [82]:
df_gb_role = df.groupby('position_role_type', as_index=False)[summable_cols].sum()
df_gb_role['spend_per_hire_USD'] = df_gb_role['spend'] / df_gb_role['count_hired']
df_gb_role.sort_values(by='spend_per_hire_USD', inplace=True, ascending=True)

In [85]:
df_gb_role

position_role_type       spend  count_started  count_submitted  \
4          WAREHOUSE  468.640259          67721            45187   
1         PRODUCTION  356.856526          51065            34304   
3           TRUCKING   49.192608           4789             2977   
0                  0   11.481912           1269              653   
2      SKILLED_TRADE   84.032345          12726             8272   

   count_approved  count_reviewed  count_expired  count_withdrawn  \
4           33486           28802           3851                9   
1           25994           24201           1614                9   
3            1792            1586            127                0   
0             463             397             51                0   
2            5280            4166            976                1   

   count_deactivated  count_position_closed  count_invited_to_interview  \
4                  0                    805                       20681   
1                  0                    143                       11175   
3                  0                     77                        1112   
0                  0                     15                         291   
2                  0                    133                        2359   

   count_hired  spend_per_hire_USD  
4          599            0.782371  
1          379            0.941574  
3           39            1.261349  
0            8            1.435239  
2           47            1.787922

## Group by position ID

In [86]:
df_gb_position = df.groupby('position_id', as_index=False)[summable_cols].sum()
df_gb_position['spend_per_hire_USD'] = df_gb_position['spend'] / df_gb_position['count_hired']
df_gb_position.sort_values(by='spend_per_hire_USD', inplace=True, ascending=True)

In [88]:
df_gb_position.head(15)

position_id     spend  count_started  count_submitted  count_approved  \
592     488966091  0.000000             78               58              42   
111     177912400  0.000000             18               18              10   
1025    738751122  0.000000             91               64              60   
1220    876731926  0.000000             14               13               8   
1146    827810278  0.002867              4                4               3   
2449   8110372407  0.003563              8                8               1   
1899   4924271866  0.005698              4                2               2   
1039    750076923  0.005991              9                8               8   
2621   8977141655  0.007150              8                5               5   
2552   8635943435  0.009663             13               10               2   
1123    816070415  0.013482              4                4               2   
1324   1005423588  0.032821              9                6               5   
813     613453387  0.016581              8                5               5   
350     334233969  0.019522             15               10               7   
954     697286556  0.019607              8                5               4   

      count_reviewed  count_expired  count_withdrawn  count_deactivated  \
592               42              0                0                  0   
111                6              4                0                  0   
1025              37             23                0                  0   
1220               7              0                0                  0   
1146               3              0                0                  0   
2449               1              0                0                  0   
1899               2              0                0                  0   
1039               8              0                0                  0   
2621               5              0                0                  0   
2552               2              0                0                  0   
1123               2              0                0                  0   
1324               5              0                0                  0   
813                5              0                0                  0   
350                7              0                0                  0   
954                3              1                0                  0   

      count_position_closed  count_invited_to_interview  count_hired  \
592                       0                          42            1   
111                       0                           6            1   
1025                      0                           9            1   
1220                      1                           7            2   
1146                      0                           1            1   
2449                      0                           1            1   
1899                      0                           2            1   
1039                      0                           8            1   
2621                      0                           3            1   
2552                      0                           2            1   
1123                      0                           2            1   
1324                      0                           5            2   
813                       0                           4            1   
350                       0                           7            1   
954                       0                           1            1   

      spend_per_hire_USD  
592             0.000000  
111             0.000000  
1025            0.000000  
1220            0.000000  
1146            0.002867  
2449            0.003563  
1899            0.005698  
1039            0.005991  
2621            0.007150  
2552            0.009663  
1123            0.013482  
1324            0.016411  
813             0.016581  
350      

In [89]:
opp_positions = df_gb_position[df_gb_position['count_hired'] > 0].sort_values(by='spend_per_hire_USD', ascending=False)
opp_positions.head(10)

position_id      spend  count_started  count_submitted  count_approved  \
509     435844256  10.037665           1608              950             569   
2186   6597822567   6.452104           1316              896             562   
920     684704823   6.289511           1294              784             682   
233     259119508   6.109201            854              561             455   
885     664024157  11.655962           1193              612             357   
2       101331110   5.786332            797              496             381   
1745   3936036304   5.688405            688              386             336   
1398   1472052002   4.808002            700              532             408   
1763   4026791137   4.026379            639              391             143   
1089    789507482   3.843678            270              136              27   

      count_reviewed  count_expired  count_withdrawn  count_deactivated  \
509              569              0                0                  0   
2186             562              0                0                  0   
920              679              0                0                  0   
233              455              0                0                  0   
885              313             43                0                  0   
2                381              0                0                  0   
1745             336              0                0                  0   
1398             408              0                0                  0   
1763             143              0                0                  0   
1089              26              0                0                  0   

      count_position_closed  count_invited_to_interview  count_hired  \
509                       0                          92            1   
2186                      0                          28            1   
920                       0                          46            1   
233                       0                         254            1   
885                       0                         162            2   
2                         0                         185            1   
1745                      0                         336            1   
1398                      0                         406            1   
1763                      0                          24            1   
1089                      0                          23            1   

      spend_per_hire_USD  
509            10.037665  
2186            6.452104  
920             6.289511  
233             6.109201  
885             5.827981  
2               5.786332  
1745            5.688405  
1398            4.808002  
1763            4.026379  
1089            3.843678

## Successes:

### Companies:

Company ID (\$ spent per hire)

1) 217743 (\$ 0.149286)
2) 629687 (\$ 0.171646)
3) 14221 (\$ 0.180456)
4) 12155 (\$ 0.199043)
5) 20805 (\$ 0.200608)
6) 27712 (\$ 0.206627)
7) 431776 (\$ 0.208275)

In [126]:
top_5_company_ids = [
    217743, 629687, 14221, 12155, 20805
]

### Role types:

Type (\$ spent per hire)

1) WAREHOUSE (\$ 0.782371)
2) PRODUCTION (\$ 0.941574)

### Positions:

Position ID (\$ spent per hire)

1) 488966091 (\$ 0.000000)
2) 177912400 (\$ 0.000000)
3) 738751122 (\$ 0.000000)
4) 876731926 (\$ 0.000000)
5) 827810278 (\$ 0.002867)
6) 8110372407 (\$ 0.003563)
7) 4924271866 (\$ 0.005698)
8) 750076923 (\$ 0.005991)
9) 8977141655 (\$ 0.007150)
10) 8635943435 (\$ 0.009663)

In [125]:
top_5_position_ids = [
    488966091, 177912400, 738751122, 876731926, 827810278
]

## Opportunities to improve:

### Companies:

Company ID (\$ spent per hire)

1) 54042 (\$ 12.626533)
2) 403553 (\$ 4.612678)
3) 346676 (\$ 3.980185)
4) 73394 (\$ 3.732378)
5) 868835 (\$ 3.561146)
6) 407848 (\$ 3.268647)
7) 58946 (\$ 3.112813)
8) 468091 (\$ 3.044571)

In [136]:
bottom_5_company_ids = [
    54042, 403553, 346676, 73394, 868835
]

### Role types:

Type (\$ spent per hire)

1) SKILLED_TRADE (\$ 1.787922)
2) TRUCKING (\$ 1.261349)

### Positions:

Position ID (\$ spent per hire)

1) 435844256 (\$ 10.037665)
2) 6597822567 (\$ 6.452104)
3) 684704823 (\$ 6.289511)
4) 259119508 (\$ 6.109201)
5) 664024157 (\$ 5.827981)
6) 101331110 (\$ 5.786332)
7) 3936036304 (\$ 5.688405)

In [135]:
bottom_5_position_ids = [
    435844256, 6597822567, 684704823, 259119508, 664024157
]

# 2. Capital spent on no-hires

In [90]:
df_no_hire = df[df['count_hired'] == 0]
df_no_hire

company_id position_role_type  position_id                      date  \
0           75621         PRODUCTION    134935318 2022-08-04 00:00:00+00:00   
1           75621         PRODUCTION   3102362893 2022-08-04 00:00:00+00:00   
2           75621         PRODUCTION   3102362893 2022-08-15 00:00:00+00:00   
3           75621         PRODUCTION   7797193831 2022-08-02 00:00:00+00:00   
4           75621         PRODUCTION   7797193831 2022-08-12 00:00:00+00:00   
...           ...                ...          ...                       ...   
40929       18738          WAREHOUSE    200792455 2022-08-10 00:00:00+00:00   
40930       18738          WAREHOUSE    200792455 2022-08-11 00:00:00+00:00   
40931       18738          WAREHOUSE    200792455 2022-08-12 00:00:00+00:00   
40932       18738          WAREHOUSE    200792455 2022-08-13 00:00:00+00:00   
40933       18738          WAREHOUSE    200792455 2022-08-14 00:00:00+00:00   

          spend  count_started  count_submitted  count_approved  \
0      0.000000              1                0               0   
1      0.003965              1                1               1   
2      0.000000              0                0               0   
3      0.000000              2                2               2   
4      0.000000              0                0               0   
...         ...            ...              ...             ...   
40929  0.032345              6                7               6   
40930  0.040191              3                3               3   
40931  0.046853              2                2               0   
40932  0.044107              1                1               1   
40933  0.044046              7                5               4   

       count_reviewed  count_expired  count_withdrawn  count_deactivated  \
0                   0              0                0                  0   
1                   0              0                0                  0   
2                   0              1                0                  0   
3                   0              0                0                  0   
4                   0              1                0                  0   
...               ...            ...              ...                ...   
40929               6              0                0                  0   
40930               3              0                0                  0   
40931               1              0                0                  0   
40932               1              0                0                  0   
40933               1              0                0                  0   

       count_position_closed  count_invited_to_interview  count_hired  
0                          0                           0            0  
1                          0                           0            0  
2                          0                           0            0  
3                          0                           0            0  
4                          0                           0            0  
...                      ...                         ...          ...  
40929                      0                           6            0  
40930                      0                           3            0  
40931                      0                           0            0  
40932                      0                           1            0  
40933                      0                           1            0  

[40062 rows x 15 columns]

In [93]:
df_no_hire_gb_position = df_no_hire.groupby('position_id', as_index=False)['spend'].sum()
total_spend_no_hire = df_no_hire_gb_position['spend'].sum()
average_spend_no_hire = df_no_hire_gb_position['spend'].mean()

In [94]:
total_spend_no_hire

950.930928866

In [95]:
average_spend_no_hire

0.35615390594232177

## Spend on no-hires

### Total for month:

\$ 950.931

### Average per role

\$ 0.356

In [163]:
df_yes_hire = df[df['count_hired'] > 0]
df_yes_hire

company_id position_role_type  position_id                      date  \
48          69615          WAREHOUSE   8049247818 2022-08-29 00:00:00+00:00   
199         69615         PRODUCTION   7146444165 2022-08-23 00:00:00+00:00   
200         69615         PRODUCTION   7146444165 2022-09-06 00:00:00+00:00   
206         69615         PRODUCTION   2146517508 2022-08-23 00:00:00+00:00   
207         69615         PRODUCTION   2146517508 2022-09-06 00:00:00+00:00   
...           ...                ...          ...                       ...   
40734      224737          WAREHOUSE    363662039 2022-09-19 00:00:00+00:00   
40873      224737         PRODUCTION    725322262 2022-09-20 00:00:00+00:00   
40907       43117         PRODUCTION   2513728158 2022-08-15 00:00:00+00:00   
40911       43117         PRODUCTION   2513728158 2022-08-22 00:00:00+00:00   
40934       18738          WAREHOUSE    200792455 2022-08-15 00:00:00+00:00   

         spend  count_started  count_submitted  count_approved  \
48     0.00000              0                0               0   
199    0.00000              0                0               0   
200    0.00000              0                0               0   
206    0.04983              1                1               1   
207    0.00000              0                0               0   
...        ...            ...              ...             ...   
40734  0.00000              0                0               0   
40873  0.00000              0                0               0   
40907  0.00000              0                0               0   
40911  0.00000              1                2               0   
40934  0.04401              3                1               1   

       count_reviewed  count_expired  count_withdrawn  count_deactivated  \
48                  0              0                0                  0   
199                 0              0                0                  0   
200                 0              0                0                  0   
206                 1              0                0                  0   
207                 0              0                0                  0   
...               ...            ...              ...                ...   
40734               0              0                0                  0   
40873               0              0                0                  0   
40907               0              0                0                  0   
40911               0              0                0                  0   
40934               4              0                0                  0   

       count_position_closed  count_invited_to_interview  count_hired  
48                         0                           0            1  
199                        0                           0            1  
200                        0                           0            1  
206                        0                           1            1  
207                        0                           0            1  
...                      ...                         ...          ...  
40734                      0                           0            2  
40873                      0                           0            1  
40907                      0                           0            1  
40911                      0                           0            1  
40934                      0                           2            1  

[873 rows x 15 columns]

In [164]:
df_yes_hire_gb_position = df_yes_hire.groupby('position_id', as_index=False)['spend'].sum()
total_spend_yes_hire = df_yes_hire_gb_position['spend'].sum()
average_spend_yes_hire = df_yes_hire_gb_position['spend'].mean()

In [165]:
total_spend_yes_hire

19.272721726300002

In [166]:
average_spend_yes_hire

0.037422760633592234

# 3. Funnel dynamics

In [101]:
df

company_id position_role_type  position_id                      date  \
0           75621         PRODUCTION    134935318 2022-08-04 00:00:00+00:00   
1           75621         PRODUCTION   3102362893 2022-08-04 00:00:00+00:00   
2           75621         PRODUCTION   3102362893 2022-08-15 00:00:00+00:00   
3           75621         PRODUCTION   7797193831 2022-08-02 00:00:00+00:00   
4           75621         PRODUCTION   7797193831 2022-08-12 00:00:00+00:00   
...           ...                ...          ...                       ...   
40930       18738          WAREHOUSE    200792455 2022-08-11 00:00:00+00:00   
40931       18738          WAREHOUSE    200792455 2022-08-12 00:00:00+00:00   
40932       18738          WAREHOUSE    200792455 2022-08-13 00:00:00+00:00   
40933       18738          WAREHOUSE    200792455 2022-08-14 00:00:00+00:00   
40934       18738          WAREHOUSE    200792455 2022-08-15 00:00:00+00:00   

          spend  count_started  count_submitted  count_approved  \
0      0.000000              1                0               0   
1      0.003965              1                1               1   
2      0.000000              0                0               0   
3      0.000000              2                2               2   
4      0.000000              0                0               0   
...         ...            ...              ...             ...   
40930  0.040191              3                3               3   
40931  0.046853              2                2               0   
40932  0.044107              1                1               1   
40933  0.044046              7                5               4   
40934  0.044010              3                1               1   

       count_reviewed  count_expired  count_withdrawn  count_deactivated  \
0                   0              0                0                  0   
1                   0              0                0                  0   
2                   0              1                0                  0   
3                   0              0                0                  0   
4                   0              1                0                  0   
...               ...            ...              ...                ...   
40930               3              0                0                  0   
40931               1              0                0                  0   
40932               1              0                0                  0   
40933               1              0                0                  0   
40934               4              0                0                  0   

       count_position_closed  count_invited_to_interview  count_hired  
0                          0                           0            0  
1                          0                           0            0  
2                          0                           0            0  
3                          0                           0            0  
4                          0                           0            0  
...                      ...                         ...          ...  
40930                      0                           3            0  
40931                      0                           0            0  
40932                      0                           1            0  
40933                      0                           1            0  
40934                      0                           2            1  

[40935 rows x 15 columns]

## Compare application expiration rates

### by position ID

In [114]:
df_gb_position = df.groupby('position_id', as_index=False)[summable_cols].sum()

In [115]:
df_gb_position

position_id     spend  count_started  count_submitted  count_approved  \
0       100286000  0.595761             64               40              36   
1       100778742  0.396039             41               27              15   
2       101331110  5.786332            797              496             381   
3       101926072  0.041716             11                8               7   
4       101998387  0.690089             42               21              12   
...           ...       ...            ...              ...             ...   
2665   9696000534  0.074183             11                6               3   
2666   9868359373  0.236643             27               21              18   
2667   9905840884  0.788126            159              100              95   
2668   9967558006  0.005686              3                3               3   
2669   9997425444  0.137776             19               14               8   

      count_reviewed  count_expired  count_withdrawn  count_deactivated  \
0                 36              0                0                  0   
1                  5             10                0                  0   
2                381              0                0                  0   
3                  7              0                0                  0   
4                 12              0                0                  0   
...              ...            ...              ...                ...   
2665               3              0                0                  0   
2666              18              0                0                  0   
2667              88              7                0                  0   
2668               3              0                0                  0   
2669               8              0                0                  0   

      count_position_closed  count_invited_to_interview  count_hired  
0                         0                           1            0  
1                         0                           5            0  
2                         0                         185            1  
3                         0                           7            1  
4                         0                           7            0  
...                     ...                         ...          ...  
2665                      0                           2            0  
2666                      0                           4            0  
2667                      0                           3            1  
2668                      0                           3            0  
2669                      0                           8            1  

[2670 rows x 12 columns]

In [117]:
df_expirations = df_gb_position.loc[:, ['position_id', 'count_approved', 'count_expired', 'count_position_closed', 'count_hired', 'spend']]
df_expirations['percent_expired'] = df_expirations['count_expired'] / df_expirations['count_approved'] * 100.0
df_expirations.sort_values('percent_expired', inplace=True, ascending=False)

In [154]:
df_expirations[(df_expirations['percent_expired'].between(50, 100, inclusive='neither')) & (df_expirations['count_hired'] == 0) & (df_expirations['count_position_closed'] == 0) & (df_expirations['count_approved'] > 0)]
# This filters for positions which had approved applications, were not closed, had applications expire, and yet no hire was made)

position_id  count_approved  count_expired  count_position_closed  \
988     715408768              25             24                      0   
870     652185488              21             20                      0   
494     420528626              14             13                      0   
2602   8886228410              13             12                      0   
898     673788967              39             36                      0   
...           ...             ...            ...                    ...   
1087    787869918              17              9                      0   
2201   6740934354              23             12                      0   
298     305498553              35             18                      0   
1966   5362240788              39             20                      0   
948     693568805              51             26                      0   

      count_hired     spend  percent_expired  
988             0  0.149843        96.000000  
870             0  0.578826        95.238095  
494             0  0.383667        92.857143  
2602            0  0.199283        92.307692  
898             0  0.345417        92.307692  
...           ...       ...              ...  
1087            0  0.437243        52.941176  
2201            0  0.167584        52.173913  
298             0  0.733940        51.428571  
1966            0  0.571444        51.282051  
948             0  0.295941        50.980392  

[132 rows x 7 columns]

* 229 positions had approved applicants, were not canceled, and yet had a 100% expiration rate

* 132 positions had approved applicants, were not canceled, did not lead to a hire, and yet had an expiration rate between 50 and 100%

In [134]:
df_expirations[df_expirations['position_id'].isin(top_5_position_ids)]

position_id  count_approved  count_expired  count_position_closed  \
111     177912400              10              4                      0   
1025    738751122              60             23                      0   
592     488966091              42              0                      0   
1146    827810278               3              0                      0   
1220    876731926               8              0                      1   

      count_hired     spend  percent_expired  
111             1  0.000000        40.000000  
1025            1  0.000000        38.333333  
592             1  0.000000         0.000000  
1146            1  0.002867         0.000000  
1220            2  0.000000         0.000000

In [133]:
df_expirations[df_expirations['position_id'].isin(top_5_position_ids)]['percent_expired'].mean()

15.666666666666668

* The top-performing 5 positions had a mean expiration percentage of 15.7%...interesting!

* Two of them had percentages of 38.3% and 40%

In [137]:
df_expirations[df_expirations['position_id'].isin(bottom_5_position_ids)]

position_id  count_approved  count_expired  count_position_closed  \
885     664024157             357             43                      0   
2186   6597822567             562              0                      0   
509     435844256             569              0                      0   
233     259119508             455              0                      0   
920     684704823             682              0                      0   

      count_hired      spend  percent_expired  
885             2  11.655962        12.044818  
2186            1   6.452104         0.000000  
509             1  10.037665         0.000000  
233             1   6.109201         0.000000  
920             1   6.289511         0.000000

In [138]:
df_expirations[df_expirations['position_id'].isin(bottom_5_position_ids)]['percent_expired'].mean()

2.4089635854341735

* It would take a bit of statistical analysis to be more certain, but it seems there is not a correlation between expiration percentage and efficient spend

### by company ID

In [139]:
df_gb_company = df.groupby('company_id', as_index=False)[summable_cols].sum()

In [140]:
df_gb_company

company_id      spend  count_started  count_submitted  count_approved  \
0         11405  19.257995           2606             1694            1154   
1         11767   0.955392            301              212             130   
2         12155   0.995217             89               57              46   
3         13213   0.303579             57               36              18   
4         13243   0.063154              5                3               3   
..          ...        ...            ...              ...             ...   
141      875301   0.432265             38               29              22   
142      884196   1.426665            157               90              61   
143      913406   6.058017            839              578             432   
144      916768   3.030271            555              396             355   
145      940803   0.309704             19                9               8   

     count_reviewed  count_expired  count_withdrawn  count_deactivated  \
0              1086             59                1                  0   
1               113             17                0                  0   
2                46              0                0                  0   
3                15              1                0                  0   
4                 3              0                0                  0   
..              ...            ...              ...                ...   
141              21              1                0                  0   
142              61              0                0                  0   
143             414             15                1                  0   
144             206            148                1                  0   
145               8              0                0                  0   

     count_position_closed  count_invited_to_interview  count_hired  
0                        7                         371           10  
1                        0                         113            1  
2                        0                          46            5  
3                        2                           9            1  
4                        0                           1            0  
..                     ...                         ...          ...  
141                      0                          12            0  
142                      0                          36            1  
143                      1                         219            9  
144                      0                          69            0  
145                      0                           8            0  

[146 rows x 12 columns]

In [143]:
df_company_expirations = df_gb_company.loc[:, ['company_id', 'count_approved', 'count_expired', 'count_position_closed', 'count_hired', 'spend']]
df_company_expirations['percent_expired'] = df_company_expirations['count_expired'] / df_company_expirations['count_approved'] * 100.0
df_company_expirations.sort_values('percent_expired', inplace=True, ascending=False)

In [148]:
df_company_expirations[(df_company_expirations['percent_expired'].between(50, 100, inclusive='right')) & (df_company_expirations['count_hired'] == 0) & (df_company_expirations['count_position_closed'] == 0) & (df_company_expirations['count_approved'] > 0)]

company_id  count_approved  count_expired  count_position_closed  \
88       294124               1              1                      0   
113      551504               8              8                      0   
86       275870               1              1                      0   
93       365654              18             18                      0   
57        75621               5              5                      0   
83       227455              12             12                      0   
41        54578               3              3                      0   
74       114903              35             32                      0   
24        34197               7              6                      0   

     count_hired     spend  percent_expired  
88             0  0.006747       100.000000  
113            0  0.032321       100.000000  
86             0  0.000049       100.000000  
93             0  0.000000       100.000000  
57             0  0.006113       100.000000  
83             0  0.259105       100.000000  
41             0  0.026953       100.000000  
74             0  0.360034        91.428571  
24             0  0.033480        85.714286

* of the 146 companies in the database, 9 showed opportunity for growth, with an expiration rate above 85% for positions where no hire was made (no other company was above 50%)

In [155]:
df_company_expirations[(df_company_expirations['percent_expired'].between(50, 100, inclusive='right')) & (df_company_expirations['count_position_closed'] == 0) & (df_company_expirations['count_approved'] > 0)]

company_id  count_approved  count_expired  count_position_closed  \
88       294124               1              1                      0   
113      551504               8              8                      0   
86       275870               1              1                      0   
93       365654              18             18                      0   
57        75621               5              5                      0   
83       227455              12             12                      0   
41        54578               3              3                      0   
74       114903              35             32                      0   
24        34197               7              6                      0   
62        78030              76             64                      0   
87       290288              57             34                      0   
58        76056              90             51                      0   

     count_hired     spend  percent_expired  
88             0  0.006747       100.000000  
113            0  0.032321       100.000000  
86             0  0.000049       100.000000  
93             0  0.000000       100.000000  
57             0  0.006113       100.000000  
83             0  0.259105       100.000000  
41             0  0.026953       100.000000  
74             0  0.360034        91.428571  
24             0  0.033480        85.714286  
62             1  0.955295        84.210526  
87             1  0.711783        59.649123  
58             2  1.447297        56.666667

* 3 others may be added to the "growth opportunity" pile when considering high expiration rates despite having made hires

In [149]:
df_company_expirations[df_company_expirations['company_id'].isin(top_5_company_ids)]

company_id  count_approved  count_expired  count_position_closed  \
118      629687            1124             38                      1   
5         14221              62              1                      4   
12        20805             145              2                      7   
2         12155              46              0                      0   
80       217743              42              0                      0   

     count_hired      spend  percent_expired  
118           85  14.589868         3.380783  
5              2   0.360912         1.612903  
12             5   1.003038         1.379310  
2              5   0.995217         0.000000  
80             3   0.447858         0.000000

In [151]:
df_company_expirations[df_company_expirations['company_id'].isin(top_5_company_ids)]['percent_expired'].mean()

1.2745992977567009

* The top-performing 5 companies had an average expiration rate of only 1.275%

* Several did have positions they closed, however

In [152]:
df_company_expirations[df_company_expirations['company_id'].isin(bottom_5_company_ids)]

company_id  count_approved  count_expired  count_position_closed  \
140      868835             198            115                     32   
54        73394             561            216                      0   
92       346676             134             37                      0   
97       403553            3189             74                    352   
39        54042             621             14                      0   

     count_hired      spend  percent_expired  
140            1   3.561146        58.080808  
54             1   3.732378        38.502674  
92             1   3.980185        27.611940  
97             6  27.676070         2.320477  
39             1  12.626533         2.254428

In [153]:
df_company_expirations[df_company_expirations['company_id'].isin(bottom_5_company_ids)]['percent_expired'].mean()

25.754065431187303

* Of the bottom-performing 5 companies, 3 allowed an expiration rate higher than 25%

* The average for those 5 is 25.75%

### by role type

In [156]:
df_gb_role = df.groupby('position_role_type', as_index=False)[summable_cols].sum()

In [158]:
df_gb_role = df_gb_role.drop(0)

In [159]:
df_gb_role

position_role_type       spend  count_started  count_submitted  \
1         PRODUCTION  356.856526          51065            34304   
2      SKILLED_TRADE   84.032345          12726             8272   
3           TRUCKING   49.192608           4789             2977   
4          WAREHOUSE  468.640259          67721            45187   

   count_approved  count_reviewed  count_expired  count_withdrawn  \
1           25994           24201           1614                9   
2            5280            4166            976                1   
3            1792            1586            127                0   
4           33486           28802           3851                9   

   count_deactivated  count_position_closed  count_invited_to_interview  \
1                  0                    143                       11175   
2                  0                    133                        2359   
3                  0                     77                        1112   
4                  0                    805                       20681   

   count_hired  
1          379  
2           47  
3           39  
4          599

In [160]:
df_role_expirations = df_gb_role.loc[:, ['position_role_type', 'count_approved', 'count_expired', 'count_position_closed', 'count_hired', 'spend']]
df_role_expirations['percent_expired'] = df_role_expirations['count_expired'] / df_role_expirations['count_approved'] * 100.0
df_role_expirations.sort_values('percent_expired', inplace=True, ascending=False)

In [162]:
df_role_expirations

position_role_type  count_approved  count_expired  count_position_closed  \
2      SKILLED_TRADE            5280            976                    133   
4          WAREHOUSE           33486           3851                    805   
3           TRUCKING            1792            127                     77   
1         PRODUCTION           25994           1614                    143   

   count_hired       spend  percent_expired  
2           47   84.032345        18.484848  
4          599  468.640259        11.500328  
3           39   49.192608         7.087054  
1          379  356.856526         6.209125

* roles classified as "skilled trades" are an area to address strategically, as they are the most expensive to hire and have the highest expiration rate

In [155]:
df_company_expirations[(df_company_expirations['percent_expired'].between(50, 100, inclusive='right')) & (df_company_expirations['count_position_closed'] == 0) & (df_company_expirations['count_approved'] > 0)]

company_id  count_approved  count_expired  count_position_closed  \
88       294124               1              1                      0   
113      551504               8              8                      0   
86       275870               1              1                      0   
93       365654              18             18                      0   
57        75621               5              5                      0   
83       227455              12             12                      0   
41        54578               3              3                      0   
74       114903              35             32                      0   
24        34197               7              6                      0   
62        78030              76             64                      0   
87       290288              57             34                      0   
58        76056              90             51                      0   

     count_hired     spend  percent_expired  
88             0  0.006747       100.000000  
113            0  0.032321       100.000000  
86             0  0.000049       100.000000  
93             0  0.000000       100.000000  
57             0  0.006113       100.000000  
83             0  0.259105       100.000000  
41             0  0.026953       100.000000  
74             0  0.360034        91.428571  
24             0  0.033480        85.714286  
62             1  0.955295        84.210526  
87             1  0.711783        59.649123  
58             2  1.447297        56.666667

* 3 others may be added to the "growth opportunity" pile when considering high expiration rates despite having made hires

In [149]:
df_company_expirations[df_company_expirations['company_id'].isin(top_5_company_ids)]

company_id  count_approved  count_expired  count_position_closed  \
118      629687            1124             38                      1   
5         14221              62              1                      4   
12        20805             145              2                      7   
2         12155              46              0                      0   
80       217743              42              0                      0   

     count_hired      spend  percent_expired  
118           85  14.589868         3.380783  
5              2   0.360912         1.612903  
12             5   1.003038         1.379310  
2              5   0.995217         0.000000  
80             3   0.447858         0.000000

In [151]:
df_company_expirations[df_company_expirations['company_id'].isin(top_5_company_ids)]['percent_expired'].mean()

1.2745992977567009

* The top-performing 5 companies had an average expiration rate of only 1.275%

* Several did have positions they closed, however

In [152]:
df_company_expirations[df_company_expirations['company_id'].isin(bottom_5_company_ids)]

company_id  count_approved  count_expired  count_position_closed  \
140      868835             198            115                     32   
54        73394             561            216                      0   
92       346676             134             37                      0   
97       403553            3189             74                    352   
39        54042             621             14                      0   

     count_hired      spend  percent_expired  
140            1   3.561146        58.080808  
54             1   3.732378        38.502674  
92             1   3.980185        27.611940  
97             6  27.676070         2.320477  
39             1  12.626533         2.254428

In [153]:
df_company_expirations[df_company_expirations['company_id'].isin(bottom_5_company_ids)]['percent_expired'].mean()

25.754065431187303

* Of the bottom-performing 5 companies, 3 allowed an expiration rate higher than 25%

* The average for those 5 is 25.75%

## Consider timeline

In [167]:
df_hired = df.loc[df['count_hired'] > 0]
df_hired

company_id position_role_type  position_id                      date  \
48          69615          WAREHOUSE   8049247818 2022-08-29 00:00:00+00:00   
199         69615         PRODUCTION   7146444165 2022-08-23 00:00:00+00:00   
200         69615         PRODUCTION   7146444165 2022-09-06 00:00:00+00:00   
206         69615         PRODUCTION   2146517508 2022-08-23 00:00:00+00:00   
207         69615         PRODUCTION   2146517508 2022-09-06 00:00:00+00:00   
...           ...                ...          ...                       ...   
40734      224737          WAREHOUSE    363662039 2022-09-19 00:00:00+00:00   
40873      224737         PRODUCTION    725322262 2022-09-20 00:00:00+00:00   
40907       43117         PRODUCTION   2513728158 2022-08-15 00:00:00+00:00   
40911       43117         PRODUCTION   2513728158 2022-08-22 00:00:00+00:00   
40934       18738          WAREHOUSE    200792455 2022-08-15 00:00:00+00:00   

         spend  count_started  count_submitted  count_approved  \
48     0.00000              0                0               0   
199    0.00000              0                0               0   
200    0.00000              0                0               0   
206    0.04983              1                1               1   
207    0.00000              0                0               0   
...        ...            ...              ...             ...   
40734  0.00000              0                0               0   
40873  0.00000              0                0               0   
40907  0.00000              0                0               0   
40911  0.00000              1                2               0   
40934  0.04401              3                1               1   

       count_reviewed  count_expired  count_withdrawn  count_deactivated  \
48                  0              0                0                  0   
199                 0              0                0                  0   
200                 0              0                0                  0   
206                 1              0                0                  0   
207                 0              0                0                  0   
...               ...            ...              ...                ...   
40734               0              0                0                  0   
40873               0              0                0                  0   
40907               0              0                0                  0   
40911               0              0                0                  0   
40934               4              0                0                  0   

       count_position_closed  count_invited_to_interview  count_hired  
48                         0                           0            1  
199                        0                           0            1  
200                        0                           0            1  
206                        0                           1            1  
207                        0                           0            1  
...                      ...                         ...          ...  
40734                      0                           0            2  
40873                      0                           0            1  
40907                      0                           0            1  
40911                      0                           0            1  
40934                      0                           2            1  

[873 rows x 15 columns]

In [177]:
hired_ids = list(df_hired['position_id'].unique())

In [178]:
df[df['position_id'].isin(hired_ids)]

company_id position_role_type  position_id                      date  \
36          69615          WAREHOUSE   8049247818 2022-08-01 00:00:00+00:00   
37          69615          WAREHOUSE   8049247818 2022-08-02 00:00:00+00:00   
38          69615          WAREHOUSE   8049247818 2022-08-03 00:00:00+00:00   
39          69615          WAREHOUSE   8049247818 2022-08-08 00:00:00+00:00   
40          69615          WAREHOUSE   8049247818 2022-08-11 00:00:00+00:00   
...           ...                ...          ...                       ...   
40930       18738          WAREHOUSE    200792455 2022-08-11 00:00:00+00:00   
40931       18738          WAREHOUSE    200792455 2022-08-12 00:00:00+00:00   
40932       18738          WAREHOUSE    200792455 2022-08-13 00:00:00+00:00   
40933       18738          WAREHOUSE    200792455 2022-08-14 00:00:00+00:00   
40934       18738          WAREHOUSE    200792455 2022-08-15 00:00:00+00:00   

          spend  count_started  count_submitted  count_approved  \
36     0.005527              2                1               1   
37     0.005857              0                0               0   
38     0.046157              1                1               1   
39     0.011225              2                1               1   
40     0.000549              1                1               1   
...         ...            ...              ...             ...   
40930  0.040191              3                3               3   
40931  0.046853              2                2               0   
40932  0.044107              1                1               1   
40933  0.044046              7                5               4   
40934  0.044010              3                1               1   

       count_reviewed  count_expired  count_withdrawn  count_deactivated  \
36                  0              0                0                  0   
37                  1              0                0                  0   
38                  0              0                0                  0   
39                  0              0                0                  0   
40                  3              0                0                  0   
...               ...            ...              ...                ...   
40930               3              0                0                  0   
40931               1              0                0                  0   
40932               1              0                0                  0   
40933               1              0                0                  0   
40934               4              0                0                  0   

       count_position_closed  count_invited_to_interview  count_hired  
36                         0                           0            0  
37                         0                           1            0  
38                         0                           0            0  
39                         0                           0            0  
40                         0                           3            0  
...                      ...                         ...          ...  
40930                      0                           3            0  
40931                      0                           0            0  
40932                      0                           1            0  
40933                      0                           1            0  
40934                      0                           2            1  

[12883 rows x 15 columns]

In [200]:
test_df = df[df['position_id'] == 8049247818]
test_df.reset_index(inplace=True, drop=True)
test_df

company_id position_role_type  position_id                      date  \
0        69615          WAREHOUSE   8049247818 2022-08-01 00:00:00+00:00   
1        69615          WAREHOUSE   8049247818 2022-08-02 00:00:00+00:00   
2        69615          WAREHOUSE   8049247818 2022-08-03 00:00:00+00:00   
3        69615          WAREHOUSE   8049247818 2022-08-08 00:00:00+00:00   
4        69615          WAREHOUSE   8049247818 2022-08-11 00:00:00+00:00   
5        69615          WAREHOUSE   8049247818 2022-08-12 00:00:00+00:00   
6        69615          WAREHOUSE   8049247818 2022-08-14 00:00:00+00:00   
7        69615          WAREHOUSE   8049247818 2022-08-15 00:00:00+00:00   
8        69615          WAREHOUSE   8049247818 2022-08-16 00:00:00+00:00   
9        69615          WAREHOUSE   8049247818 2022-08-18 00:00:00+00:00   
10       69615          WAREHOUSE   8049247818 2022-08-21 00:00:00+00:00   
11       69615          WAREHOUSE   8049247818 2022-08-22 00:00:00+00:00   
12       69615          WAREHOUSE   8049247818 2022-08-29 00:00:00+00:00   

       spend  count_started  count_submitted  count_approved  count_reviewed  \
0   0.005527              2                1               1               0   
1   0.005857              0                0               0               1   
2   0.046157              1                1               1               0   
3   0.011225              2                1               1               0   
4   0.000549              1                1               1               3   
5   0.008016              1                0               0               0   
6   0.011640              1                2               2               0   
7   0.023024              3                0               0               0   
8   0.023866              1                1               1               2   
9   0.004319              1                1               1               0   
10  0.004905              1                0               0               0   
11  0.016032              1                1               0               2   
12  0.000000              0                0               0               0   

    count_expired  count_withdrawn  count_deactivated  count_position_closed  \
0               0                0                  0                      0   
1               0                0                  0                      0   
2               0                0                  0                      0   
3               0                0                  0                      0   
4               0                0                  0                      0   
5               0                0                  0                      0   
6               0                0                  0                      0   
7               0                0                  0                      0   
8               0                0                  0                      0   
9               0                0                  0                      0   
10              0                0                  0                      0   
11              0                0                  0                      0   
12              0                0                  0                      0   

    count_invited_to_interview  count_hired  
0                            0            0  
1                            1            0  
2                            0            0  
3                            0            0  
4                            3            0  
5                            0            0  
6                            0            0  
7                            0            0  
8                            1            0  
9                            0            0  
10                           0            0  
11                           1            0  
12                           0            1

In [205]:
value = test_df.loc[0, 'count_approved'] - test_df.loc[0, 'count_reviewed']
test_df.loc[0, 'backlog'] = int(value)
test_df

company_id position_role_type  position_id                      date  \
0        69615          WAREHOUSE   8049247818 2022-08-01 00:00:00+00:00   
1        69615          WAREHOUSE   8049247818 2022-08-02 00:00:00+00:00   
2        69615          WAREHOUSE   8049247818 2022-08-03 00:00:00+00:00   
3        69615          WAREHOUSE   8049247818 2022-08-08 00:00:00+00:00   
4        69615          WAREHOUSE   8049247818 2022-08-11 00:00:00+00:00   
5        69615          WAREHOUSE   8049247818 2022-08-12 00:00:00+00:00   
6        69615          WAREHOUSE   8049247818 2022-08-14 00:00:00+00:00   
7        69615          WAREHOUSE   8049247818 2022-08-15 00:00:00+00:00   
8        69615          WAREHOUSE   8049247818 2022-08-16 00:00:00+00:00   
9        69615          WAREHOUSE   8049247818 2022-08-18 00:00:00+00:00   
10       69615          WAREHOUSE   8049247818 2022-08-21 00:00:00+00:00   
11       69615          WAREHOUSE   8049247818 2022-08-22 00:00:00+00:00   
12       69615          WAREHOUSE   8049247818 2022-08-29 00:00:00+00:00   

       spend  count_started  count_submitted  count_approved  count_reviewed  \
0   0.005527              2                1               1               0   
1   0.005857              0                0               0               1   
2   0.046157              1                1               1               0   
3   0.011225              2                1               1               0   
4   0.000549              1                1               1               3   
5   0.008016              1                0               0               0   
6   0.011640              1                2               2               0   
7   0.023024              3                0               0               0   
8   0.023866              1                1               1               2   
9   0.004319              1                1               1               0   
10  0.004905              1                0               0               0   
11  0.016032              1                1               0               2   
12  0.000000              0                0               0               0   

    count_expired  count_withdrawn  count_deactivated  count_position_closed  \
0               0                0                  0                      0   
1               0                0                  0                      0   
2               0                0                  0                      0   
3               0                0                  0                      0   
4               0                0                  0                      0   
5               0                0                  0                      0   
6               0                0                  0                      0   
7               0                0                  0                      0   
8               0                0                  0                      0   
9               0                0                  0                      0   
10              0                0                  0                      0   
11              0                0                  0                      0   
12              0                0                  0                      0   

    count_invited_to_interview  count_hired  backlog  
0                            0            0      1.0  
1                            1            0      NaN  
2                            0            0      NaN  
3                            0            0      NaN  
4                            3            0      NaN  
5                            0            0      NaN  
6                            0            0      NaN  
7                            0            0      NaN  
8                            1            0      NaN  
9                            0            0      NaN  
10                           0            0      NaN  
11                           1       

In [206]:
len(test_df)

13

In [208]:
for i in range(1, len(test_df)):
    current_backlog = test_df.loc[i-1, 'backlog']
    new_backlog = current_backlog - test_df.loc[i, 'count_reviewed'] - test_df.loc[i, 'count_expired'] - test_df.loc[i, 'count_withdrawn'] - test_df.loc[i, 'count_deactivated'] - test_df.loc[i, 'count_position_closed'] + test_df.loc[i, 'count_approved']
    test_df.loc[i, 'backlog'] = new_backlog

In [209]:
test_df

company_id position_role_type  position_id                      date  \
0        69615          WAREHOUSE   8049247818 2022-08-01 00:00:00+00:00   
1        69615          WAREHOUSE   8049247818 2022-08-02 00:00:00+00:00   
2        69615          WAREHOUSE   8049247818 2022-08-03 00:00:00+00:00   
3        69615          WAREHOUSE   8049247818 2022-08-08 00:00:00+00:00   
4        69615          WAREHOUSE   8049247818 2022-08-11 00:00:00+00:00   
5        69615          WAREHOUSE   8049247818 2022-08-12 00:00:00+00:00   
6        69615          WAREHOUSE   8049247818 2022-08-14 00:00:00+00:00   
7        69615          WAREHOUSE   8049247818 2022-08-15 00:00:00+00:00   
8        69615          WAREHOUSE   8049247818 2022-08-16 00:00:00+00:00   
9        69615          WAREHOUSE   8049247818 2022-08-18 00:00:00+00:00   
10       69615          WAREHOUSE   8049247818 2022-08-21 00:00:00+00:00   
11       69615          WAREHOUSE   8049247818 2022-08-22 00:00:00+00:00   
12       69615          WAREHOUSE   8049247818 2022-08-29 00:00:00+00:00   

       spend  count_started  count_submitted  count_approved  count_reviewed  \
0   0.005527              2                1               1               0   
1   0.005857              0                0               0               1   
2   0.046157              1                1               1               0   
3   0.011225              2                1               1               0   
4   0.000549              1                1               1               3   
5   0.008016              1                0               0               0   
6   0.011640              1                2               2               0   
7   0.023024              3                0               0               0   
8   0.023866              1                1               1               2   
9   0.004319              1                1               1               0   
10  0.004905              1                0               0               0   
11  0.016032              1                1               0               2   
12  0.000000              0                0               0               0   

    count_expired  count_withdrawn  count_deactivated  count_position_closed  \
0               0                0                  0                      0   
1               0                0                  0                      0   
2               0                0                  0                      0   
3               0                0                  0                      0   
4               0                0                  0                      0   
5               0                0                  0                      0   
6               0                0                  0                      0   
7               0                0                  0                      0   
8               0                0                  0                      0   
9               0                0                  0                      0   
10              0                0                  0                      0   
11              0                0                  0                      0   
12              0                0                  0                      0   

    count_invited_to_interview  count_hired  backlog  
0                            0            0      1.0  
1                            1            0      0.0  
2                            0            0      1.0  
3                            0            0      2.0  
4                            3            0      0.0  
5                            0            0      0.0  
6                            0            0      2.0  
7                            0            0      2.0  
8                            1            0      1.0  
9                            0            0      2.0  
10                           0            0      2.0  
11                           1       

In [229]:
last_backlog_index = 0
for i in range(1, len(test_df)):
    print(f'In outer for loop')
    print(f'Index: {i}')
    print(f'Last backlog index is now {last_backlog_index}')
    day = test_df.loc[i, 'date'].day
    print(f'Day: {day}')

    if test_df.loc[i, 'backlog'] > 0 and (test_df.loc[i-1, 'backlog'] == 0):
        last_backlog_index = i
        print(f'Last backlog index is now {last_backlog_index}')
        print(f'Backlog in row {i} is greater than 0, continuing.')
        # print('\n*******\n')
        continue
    elif test_df.loc[i, 'backlog'] > 0 and (test_df.loc[i-1, 'backlog'] > 0):
        print(f'Last backlog index is still {last_backlog_index}')
        print(f'Backlog in row {i} is greater than 0, continuing.')
        # print('\n*******\n')
        continue
    else:
        print(f'Last backlog index: {last_backlog_index}')
        for j in range(i-1, last_backlog_index-1, -1):
            print(f'In inner for loop')
            print(f'Checking row {j}')
            if test_df.loc[j, 'backlog'] > 0:
                delta = day - test_df.loc[j, 'date'].day
                print(f'Delta: {delta}')
                test_df.loc[i, 'time_to_clear'] = delta
                # last_backlog_index = j
            else:
                print(f'Backlog in row {j} is 0.')
                last_backlog_index = i
                print(f'Last backlog index is now {last_backlog_index}')

    print('\n*******\n')

In outer for loop
Index: 1
Last backlog index is now 0
Day: 2
Last backlog index: 0
In inner for loop
Checking row 0
Delta: 1

*******

In outer for loop
Index: 2
Last backlog index is now 0
Day: 3
Last backlog index is now 2
Backlog in row 2 is greater than 0, continuing.
In outer for loop
Index: 3
Last backlog index is now 2
Day: 8
Last backlog index is still 2
Backlog in row 3 is greater than 0, continuing.
In outer for loop
Index: 4
Last backlog index is now 2
Day: 11
Last backlog index: 2
In inner for loop
Checking row 3
Delta: 3
In inner for loop
Checking row 2
Delta: 8

*******

In outer for loop
Index: 5
Last backlog index is now 2
Day: 12
Last backlog index: 2
In inner for loop
Checking row 4
Backlog in row 4 is 0.
Last backlog index is now 5
In inner for loop
Checking row 3
Delta: 4
In inner for loop
Checking row 2
Delta: 9

*******

In outer for loop
Index: 6
Last backlog index is now 5
Day: 14
Last backlog index is now 6
Backlog in row 6 is greater than 0, continuing.
In ou

In [223]:
test_df

company_id position_role_type  position_id                      date  \
0        69615          WAREHOUSE   8049247818 2022-08-01 00:00:00+00:00   
1        69615          WAREHOUSE   8049247818 2022-08-02 00:00:00+00:00   
2        69615          WAREHOUSE   8049247818 2022-08-03 00:00:00+00:00   
3        69615          WAREHOUSE   8049247818 2022-08-08 00:00:00+00:00   
4        69615          WAREHOUSE   8049247818 2022-08-11 00:00:00+00:00   
5        69615          WAREHOUSE   8049247818 2022-08-12 00:00:00+00:00   
6        69615          WAREHOUSE   8049247818 2022-08-14 00:00:00+00:00   
7        69615          WAREHOUSE   8049247818 2022-08-15 00:00:00+00:00   
8        69615          WAREHOUSE   8049247818 2022-08-16 00:00:00+00:00   
9        69615          WAREHOUSE   8049247818 2022-08-18 00:00:00+00:00   
10       69615          WAREHOUSE   8049247818 2022-08-21 00:00:00+00:00   
11       69615          WAREHOUSE   8049247818 2022-08-22 00:00:00+00:00   
12       69615          WAREHOUSE   8049247818 2022-08-29 00:00:00+00:00   

       spend  count_started  count_submitted  count_approved  count_reviewed  \
0   0.005527              2                1               1               0   
1   0.005857              0                0               0               1   
2   0.046157              1                1               1               0   
3   0.011225              2                1               1               0   
4   0.000549              1                1               1               3   
5   0.008016              1                0               0               0   
6   0.011640              1                2               2               0   
7   0.023024              3                0               0               0   
8   0.023866              1                1               1               2   
9   0.004319              1                1               1               0   
10  0.004905              1                0               0               0   
11  0.016032              1                1               0               2   
12  0.000000              0                0               0               0   

    count_expired  count_withdrawn  count_deactivated  count_position_closed  \
0               0                0                  0                      0   
1               0                0                  0                      0   
2               0                0                  0                      0   
3               0                0                  0                      0   
4               0                0                  0                      0   
5               0                0                  0                      0   
6               0                0                  0                      0   
7               0                0                  0                      0   
8               0                0                  0                      0   
9               0                0                  0                      0   
10              0                0                  0                      0   
11              0                0                  0                      0   
12              0                0                  0                      0   

    count_invited_to_interview  count_hired  backlog  time_to_clear  
0                            0            0      1.0            NaN  
1                            1            0      0.0            1.0  
2                            0            0      1.0            NaN  
3                            0            0      2.0            NaN  
4                            3            0      0.0           10.0  
5                            0            0      0.0           11.0  
6                            0            0      2.0            NaN  
7                            0            0      2.0            NaN  
8                            1            0      1.0            NaN

In [212]:
for i in range(0, -1, -1):
    print(i)

0
